src/Castle.Core/DynamicProxy/IInvocation.cs

```diff
@@ -114,6 +114,12 @@ public interface IInvocation
 		/// </remarks>
 		void Proceed();
 
+		/// <summary>
+		///   Returns an object describing the <see cref="Proceed"/> operation for this <see cref="IInvocation"/>
+		///   at this specific point during interception.
+		/// </summary>
+		IInvocationProceedInfo GetProceedInfo();
```

Comment:
Something is bugging me that this looks too normal, like a part of the API people should actually use commonly. Like how no one thinks twice about using `Type.GetTypeInfo` even though many of the methods have been added back to `Type` in .NET Standard.

I wondered why we couldn't just have a `CreateProceedDelegate` (or `GetProceedDelegate`) and remembered you said this about allocations:

> This could be mitigated by simply having InvocationProceedInfo have a Action proceed field instead of a combination of IInvocation + int. But then someone would have to allocate that delegate, which also creates an object on the heap. Nothing gained!

This is currently allocating the `ProcessInfo` object, what if we just allocated a **delegate instead**? Could we have a private `Proceed(int index)` (or maybe `ProceedCompleted(int)` which throws if the invocation isn't flagged completed yet) that `CreateProceedDelegate` would call? The `Proceed` `MethodInfo` is already cached, we can easily do this one too.

Reply:
There's no way to *just* allocate a delegate. You'll need at least one additional object that holds the copied interceptor index. Say if you returned a lambda that captures the `currentInterceptorIndex`, the C# compiler will transparently copy that over into a display class which will act as the delegate's `Target` object. So there'll be two objects, vs. just the one in the case of `IInvocationProceedInfo`.

Adding a `private void Proceed(int index)` won't help because `CreateProceedDelegate` should return a parameterless `Action` (lest we expose `AbstractInvocation`'s internals!), but that delegate will still have to know what argument to pass to `Proceed(int)`. So we're back to needing an additional state object.

Reply:
> Something is bugging me that this looks too normal, like a part of the API people should actually use commonly.

I think I know what you mean. I think there's a trade-off to be made here: #428 (i.e. giving `Intercept` an additional parameter `Action proceed`) is elegant but a big breaking change. This PR is less elegant but a much smaller breaking change, and more efficient. `Action CreateProceedDelegate` is somewhere in-between, also from an efficiency perspective.

Reply:
> There's no way to _just_ allocate a delegate. You'll need at least one additional object that holds the copied interceptor index

Of course the compiler's capture class, I'm such an idiot. Sorry for my sloppy review, I shouldn't do code reviews at 1am 😞.

I'm now coming around to this idea, maybe just some small naming changes, you've definitely thought about it a lot more than me.

I guess my next question is do you see `IInvocationProceedInfo` a read-only info class (like `TypeInfo`) or something more stateful like `System.Threading.ExecutionContext`? My guess is the former since you've got `GetProceedInfo` rather than "capture" or another verb. I just wonder if the API reads clear enough that the proceed info you get is tied directly to the invocation it was created/captured from, not just any invocation of that method, and that "invoking" the proceed info actually modifies the invocation.

Would you ever consider enhancing `ProceedInfo.Invoke` to restore not just the interceptor index but also the return value as it exits in the finally block?

Maybe the naming is clear enough, the framework has `ExceptionDispatchInfo`, with static `Capture(ex)` and instance `Throw`. Had you considered hiding this functionality inside the info class without an interface, e.g. `InvocationProceedInfo.Capture(invocation)` and then `proceedInfo.Invoke`. Capture would just throw if it wasn't given an `AbstractInvocation`, not that DP works with other invocations, and users won't even know about this type unless they need it. It would move this more advanced functionality and a little bit of a wart to the side while being a supported API. Maybe I worrying way too much, anyone using DP has to have a decent idea of what they are doing.

I think at minimum we should make the XML documentation describe that this isn't a general purpose method, we don't want people doing `invocation.GetProceedInfo().Invoke()` not realising they should be using `invocation.Proceed()`.

My comments are a bit rambling, but I thought I'd keep them even though they don't provide actionable feedback. Happy to talk on anything that you think is useful, but there is no need to respond to each thing I've said in this comment.

Reply:
> Sorry for my sloppy review

Not at all, I'm glad you're approaching this from different angles than I did!

> do you see `IInvocationProceedInfo` a read-only info class

Not just that, but immutable. That's important to guarantee the fact that it captures the `Proceed` action *at a specific moment in time*.

Btw. I like the notion of "capturing" / getting a "snapshot". It would underline the immutable nature of `IInvocationProceedInfo` better! `CaptureProceedInfo`, perhaps?

I'm somewhat unsure about the best way to remove `GetProceedInfo` out of the direct sight of regular users (also not sure whether it's even necessary, users should read the docs instead of blindly using APIs IMO). If `EditorBrowsableState.Advanced` and XML docs (my preferred approach) are not sufficient, then `InvocationProceedInfo.Capture(...).Invoke()` would be a nice solution too, despite its limitations, i.e.: turning `IInvocation` into a leaky abstraction, and not being as extensible with reflection-like methods (since the info type would then be more of an opaque token like `ExceptionDispatchInfo`, rather than a reflection `...Info`).

> Would you ever consider enhancing `ProceedInfo.Invoke` to restore not just the interceptor index but also the return value

Probably not, because the return value isn't really part of the "proceed state"... and you still need the new `ReturnValue` after the `Invole` call, if you need it at all. I think we should leave cleaning up `ReturnValue` the responsibility of downstream users.

If the capture/restore became the focus then "invocation mementos" (like e.g. the one I tried out in #438) might be a more appropriate abstraction overall.

Reply:
> Not at all, I'm glad you're approaching this from different angles than I did!

👍

> Btw. I like the notion of "capturing" / getting a "snapshot". It would underline the immutable nature of `IInvocationProceedInfo` better! `CaptureProceedInfo`, perhaps?

Great, I think renaming `GetProceedInfo` to `CaptureProceedInfo` makes it clear you are taking a snapshot at that point in time, if you call capture again on the same `IInvocation` especially in another interceptor you'll obviously get a new `IInvocationProceedInfo` instance.

> I'm somewhat unsure about the best way to remove `GetProceedInfo` out of the direct sight of regular users (also not sure whether it's even necessary, users should read the docs instead of blindly using APIs IMO).

Yer, my idea of a static method is overkill, I think naming it `CaptureProceedInfo` would be enough to push someone to look at the XML docs or full docs to know what this method is for, its name doesn't sound like a normal getter that always returns the same object.

Reply:
Done.